File: backend/src/main/java/edu/cit/lemslite/Service/JwtService.java
package edu.cit.lemslite.Service;

import java.util.Base64;
import java.util.Date;
import java.util.HashMap;
import java.util.Map;
import java.util.function.Function;

import javax.crypto.SecretKey;

import org.springframework.beans.factory.annotation.Value;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.stereotype.Service;

import io.jsonwebtoken.Claims;
import io.jsonwebtoken.Jwts;
import io.jsonwebtoken.io.Decoders;
import io.jsonwebtoken.security.Keys;

@Service
public class JwtService {

	@Value("${jwt.secret}")
	private String secretKey;

	public String generateToken(String insti_id, int roleId, String first_name, String last_name, int uid) {
		Map<String, Object> claims = new HashMap<>();

		claims.put("role_id", roleId);
		claims.put("first_name", first_name);
		claims.put("last_name", last_name);
		claims.put("full_name", first_name + " " + last_name);
		claims.put("uid", uid);

		return Jwts.builder()
				.setClaims(claims)
				.setSubject(insti_id)
				.setIssuedAt(new Date(System.currentTimeMillis()))
				.setExpiration(new Date(System.currentTimeMillis() + 8 * 60 * 60 * 1000)) // 8 hours
				.signWith(getKey())
				.compact();
	}

	private SecretKey getKey() {
		byte[] keyBytes = Decoders.BASE64.decode(secretKey);
		return Keys.hmacShaKeyFor(keyBytes);
	}

	public String extractInstiId(String token) {
		return extractClaim(token, Claims::getSubject);
	}

	private <T> T extractClaim(String token, Function<Claims, T> claimResolver) {
		final Claims claims = extractAllClaims(token);
		return claimResolver.apply(claims);
	}

	private Claims extractAllClaims(String token) {
		try {
			return Jwts.parser()
					.setSigningKey(getKey())
					.build()
					.parseClaimsJws(token)
					.getBody();
		} catch (Exception e) {
			// Log the error (consider using a logger)
			System.err.println("Error decoding token: " + e.getMessage());
			return null; // or throw a custom exception
		}
	}

	public boolean isTokenExpired(String token) {
		return extractExpirationToken(token).before(new Date());
	}

	public Date extractExpirationToken(String token) {
		return extractClaim(token, Claims::getExpiration);
	}

	public boolean validateToken(String token, UserDetails userDetails) {
		Claims claims = extractAllClaims(token);
		if (claims == null) {
			return false; // Token is invalid
		}
		final String insti_id = claims.getSubject();

		return (insti_id.equals(userDetails.getUsername()) && !isTokenExpired(token));
	}
}

File: frontend_mobile/app/src/main/java/com/example/lemslite/JwtService.kt
import android.util.Base64
import android.util.Log
import io.jsonwebtoken.Claims
import io.jsonwebtoken.Jwts
import io.jsonwebtoken.security.Keys
import java.util.Date
import javax.crypto.SecretKey

class JwtService {

    // Decode the Base64-encoded secret key
    private val secretKey: SecretKey = Keys.hmacShaKeyFor(
        Base64.decode("TEVNUy1MSVRFU1lTSU5URUcyMDI1MjYyNjI2MjYyNjI2MjY=", Base64.DEFAULT)
    )

    fun isTokenExpired(token: String): Boolean {
        return try {
            val claims = extractAllClaims(token)
            val expiration = claims.expiration
            expiration.before(Date())
        } catch (e: Exception) {
            Log.e("JwtService", "Error validating token: ${e.message}", e)
            true
        }
    }

    fun extractAllClaims(token: String): Claims {
        return Jwts.parserBuilder()
            .setSigningKey(secretKey)
            .build()
            .parseClaimsJws(token)
            .body
    }

    fun getRoleIdFromToken(token: String): Int? {
        return try {
            val claims = extractAllClaims(token)
            claims["role_id"]?.toString()?.toIntOrNull()
        } catch (e: Exception) {
            Log.e("JwtService", "Error extracting role_id: ${e.message}", e)
            null
        }
    }
}